src/driver-sinowealth.c

```diff
@@ -0,0 +1,400 @@
+{
+	return (struct ratbag_color) {.red = raw.r, .green = raw.g, .blue = raw.b};
+}
+
+static RGB8
+sinowealth_color_to_raw(struct ratbag_color color)
+{
+	return (RGB8) {.r = color.red, .g = color.green, .b = color.blue};
+}
+
+static int
+sinowealth_read_profile(struct ratbag_profile *profile)
+{
+	struct ratbag_device *device = profile->device;
+	struct sinowealth_data *drv_data = device->drv_data;
+	struct sinowealth_config_report *config = &drv_data->config;
+	struct ratbag_resolution *resolution;
+	struct ratbag_led *led;
+	int num_dpis = (SINOWEALTH_DPI_MAX - SINOWEALTH_DPI_MIN) / SINOWEALTH_DPI_STEP + 1 + 1;
+	unsigned int dpis[num_dpis];
+	unsigned int hz = 1000; /* TODO */
+	int rc;
+	uint8_t *data;
+
+	uint8_t cmd[6] = {SINOWEALTH_REPORT_ID_CMD, SINOWEALTH_CMD_GET_CONFIG};
+	rc = ratbag_hidraw_set_feature_report(device, SINOWEALTH_REPORT_ID_CMD, cmd, sizeof(cmd));
+	if(rc != sizeof(cmd)) {
+		log_error(device->ratbag, "Error while sending read config command: %d\n", rc);
+		return -1;
+	}
+
+	data = zalloc(SINOWEALTH_CONFIG_SIZE);
+	rc = ratbag_hidraw_get_feature_report(device, SINOWEALTH_REPORT_ID_CONFIG,
+					      data, SINOWEALTH_CONFIG_SIZE);
+	/* The GET_FEATURE report length has to be 520, but the actual data returned is less */
+	if (rc < sizeof(*config)) {
+		log_error(device->ratbag, "Could not read device configuration: %d\n", rc);
+		return -1;
+	}
+
+	memcpy(config, data, sizeof(*config));
+	free(data);
+
+	/* TODO */
+	ratbag_profile_set_report_rate_list(profile, &hz, 1);
+	ratbag_profile_set_report_rate(profile, hz);
+
+	/* Generate DPI list */
+	dpis[0] = 0; /* 0 DPI = disabled */
+	for(int i = 1; i < num_dpis; i++) {
+		dpis[i] = SINOWEALTH_DPI_MIN + i * SINOWEALTH_DPI_STEP;
+	}
+
+	ratbag_profile_for_each_resolution(profile, resolution) {
+		if(config->config & SINOWEALTH_XY_INDEPENDENT) {
+			resolution->dpi_x = sinowealth_raw_to_dpi(config->dpi[resolution->index * 2]);
+			resolution->dpi_y = sinowealth_raw_to_dpi(config->dpi[resolution->index * 2 + 1]);
+		} else {
+			resolution->dpi_x = sinowealth_raw_to_dpi(config->dpi[resolution->index]);
+			resolution->dpi_y = resolution->dpi_x;
+		}
+		if (config->dpi_enabled & (1<<resolution->index)) {
+			/* DPI step is disabled, fake it by setting DPI to 0 */
+			resolution->dpi_x = 0;
+			resolution->dpi_y = 0;
+		}
+		resolution->is_active = resolution->index == config->active_dpi - 1;
+		resolution->is_default = resolution->is_active;
+		ratbag_resolution_set_dpi_list(resolution, dpis, num_dpis);
+		ratbag_resolution_set_cap(resolution, RATBAG_RESOLUTION_CAP_SEPARATE_XY_RESOLUTION);
+	}
+
+	/* Body lighting */
+	led = ratbag_profile_get_led(profile, 0);
+	led->type = RATBAG_LED_TYPE_SIDE;
+	led->colordepth = RATBAG_LED_COLORDEPTH_RGB_888;
+	ratbag_led_set_mode_capability(led, RATBAG_LED_OFF);
+	ratbag_led_set_mode_capability(led, RATBAG_LED_ON);
+	ratbag_led_set_mode_capability(led, RATBAG_LED_CYCLE);
+	ratbag_led_set_mode_capability(led, RATBAG_LED_BREATHING);
```

Comment:
I don't think this should be done in `read_profile`, it should probably be in `probe`.

Or do you have dynamic capabilities? I don't think piper is ready to handle that.

Reply:
This function only gets called from probe. I understood the comments about ratbag_driver to mean that a new driver is only supposed to implement probe/commit/remove, since everything else is deprecated.

Reply:
Right, sorry. I still think it would be maybe better to have the capability setting logic unbundled from `read_profile` but that's just a nitpick.

Reply:
Ah, ok. I split read_profile up and moved all the invariant stuff to a separate function (sinowealth_init_profile).